docs/internals-mappings.md

```diff
@@ -0,0 +1,428 @@
+
+An example of a Direct Assignment Mapping can be seen below:
+
+> **Note:** A more complete example can be seen under `Examples` below.
+
+```json
+{
+  "fieldMappings": [
+    {
+      "type": "DirectAssignment",
+      "directAssignment": {
+        "schemaModelName": "SolarSystemResourceSchema",
+        "schemaFieldPath": "Name",
+        "sdkModelName": "SolarSystem",
+        "sdkFieldPath": "Name"
+      }
+    }
+  ]
+}
+```
+
+### b. `ModelToModel` Field Mappings
+
+ModelToModel Field Mappings - not to be confused with ModelToModel Mappings (below) - allow mapping a Terraform Schema Model onto a field within an SDK Model.
+
+The best example of this is allowing the top-level Terraform Schema Model for a given Terraform Resource to map onto the `properties` object that exists in most payloads in Azure Resource Manager - allowing for fields which exist at the root of the Terraform Schema to be mapped to a nested object.
+
+Each ModelToModel Field Mapping must have an associated ModelToModel mapping, these are used together:
+
+* A ModelToModel Field Mapping is used to output the call to the mapping function itself.
+* A ModelToModel Mapping is used to output the mapping function, the fields which should be mapped are then inferred from the ModelToModel Field Mappings.
+
+> **Note:** A complete end-to-end example can be found below under `Example`.
+
+Model to Model Field Mappings support the following properties:
+
+* `schemaModelName` - specifies the name of the Terraform Schema Model that should be mapped onto the Field within the SDK Model.
+* `sdkModelName` - specifies the name of the SDK Model containing the SDK Field that the Terraform Schema Model should be mapped onto.
+* `sdkFieldPath` - specifies the Path to the Field within the SDK Model where the Terraform Schema Model should be mapped onto.
+
+An example of a Model to Model Field Mapping can be seen below: 
+
+```json
+{
+  "fieldMappings": [
+    {
+      "type": "ModelToModel",
+      "modelToModel": {
+        "schemaModelName": "SolarSystemResourceSchema",
+        "sdkModelName": "SolarSystem",
+        "sdkFieldName": "Properties"
+      }
+    }
+  ]
+}
+```
+
+This would allow the Terraform Generator to output:
+
+```go
+func (r VirtualMachineResource) mapVirtualMachineResourceSchemaToVirtualMachineProperties(input VirtualMachineResourceSchema, output *virtualmachines.VirtualMachine) error {
+    if output.Properties == nil {
+        output.Properties = &virtualmachines.VirtualMachineProperties{}
+    }
+    if err := r.mapVirtualMachineResourceSchemaToVirtualMachineProperties(input, output.Properties); err != nil {
+        return fmt.Errorf("mapping Schema to SDK Field %q / Model %q: %+v", "Properties", "SolarSystemProperties", err)
+    }
+    
+    return nil
+}
```

Comment:
In this example this function would be calling itself, is that what it's supposed to do?